contrib/node/src/python/pants/contrib/node/subsystems/resolvers/npm_resolver.py

```diff
@@ -35,42 +36,31 @@ def resolve_target(self, node_task, target, results_dir, node_paths):
       if not os.path.exists('package.json'):
         raise TaskError(
           'Cannot find package.json. Did you forget to put it in target sources?')
-      package_manager = node_task.get_package_manager_for_target(target=target)
       install_optional = self.get_options().install_optional
-      if package_manager == node_task.node_distribution.PACKAGE_MANAGER_NPM:
+      # TODO(rjiang): Figure out how to support source level dependencies in BUILD file.
+      package_manager = node_task.get_package_manager(target=target)
+      if package_manager == PACKAGE_MANAGER_NPM:
```

Comment:
I think it would be cleaner solution to move these pre-install steps into the package manager. Something like
package_manager = node_task.get_package_manager_util(target=target)
package_manager.preinstall(...)
result, command = node_task.install()

Reply:
That's a good idea.  I decided not to touch too much of this package.json gen codes as it will open another can of worm which will make this already big change unmanageable.  I will add a TODO here.